File: src/test/resources/variablesExtended/ScopeVariables.java
package variablesExtended;

import java.lang.RuntimeException;
import java.lang.Exception;

public class ScopeVariables {

    // The variables have mostly the same name to give the VariableUsageResolver a possibly hard time and uncover more faults.
    public String varName = "instance_field";
    public static String staticVarName = "static_field";

    public static void main(String[] args) {
        ScopeVariables scopeVariable = new ScopeVariables();
        String varName = "local_varName";
        scopeVariable.function1();
        scopeVariable.function2(varName);
        scopeVariable.function3(varName);
        scopeVariable.function4(varName);
        InnerClass innerClass = scopeVariable.new InnerClass();
        innerClass.function1();
        innerClass.function2(varName);
    }

    public void function1() {
        // Access field
        for (String varName = "first_loop_local"; varName.length() < 17; varName += " ") {
            printLog("func1_first_loop_varName", varName); // Should reference their respective loop scope
        }
        for (String varName = "second_loop_local"; varName.length() < 18; varName += " ") {
            printLog(
                    "func1_second_loop_varName", varName); // Should reference their respective loop scope
        }
        printLog("func1_imp_this_varName", varName);
    }

    public void function2(String varName) {
        printLog("func2_param_varName", varName);
        // Consumer accessOuterVarname =
        //    (lambdaVarName) -> printLog("func2_inLambda_param_varName", varName);
        // TODO implement test functions from Anonymous functions and lambdaExpressions
    }

    public void function3(String varName) {
        String staticVarName; // To try to shadow static varname in case the resolution behaves wrongly,
        // this should not shadow
        ExternalClass externalClass = new ExternalClass();
        printLog("func3_external_instance_varName", externalClass.varName);
        printLog(
                "func3_external_static_staticVarName",
                ExternalClass
                        .staticVarName); // Should reference static var in class and not be shadowed by this
        // class field
    }

    public void function4(String varName) {
        printLog(
                "func4_external_static_staticVarName",
                ExternalClass
                        .staticVarName); // Should reference static var in class and not be shadowed by this
        // class field
    }

    public class InnerClass {
        public String varName = "inner_instance_field";
        public String staticVarName =
                "inner_static_field"; // Not really a static variable but we name it like this to see if
        // variable resolution works correctly

        public void function1() {
            printLog("func1_inner_imp_this_varName", varName);
            printLog("func1_outer_this_varName", ScopeVariables.this.varName);
            printLog("func1_outer_static_staticVarName", ScopeVariables.staticVarName);
        }

        public void function2(String varName) {
            printLog("func2_inner_param_varName", varName);
            printLog("func2_inner_this_varName", this.varName);
        }

        public void function3(){
            try{
                throw new RuntimeException("excpetion");
            }catch(RuntimeException staticVarName){
                printLog("func3_inner_exception_staticVarName", staticVarName);
            }catch(Exception varName){
                printLog("func3_inner_exception_varName", varName);
            }
        }
    }

    public static void printLog(String logID, String message) {
        System.out.println(logID + ": " + message);
    }
    public static void printLog(String logID, Exception message) {
        System.out.println(logID + ": " + message.getMessage());
    }
}

File: cpg-language-cxx/src/test/resources/variables_extended/cpp/scope_variables.cpp
#include<iostream>
#include<string>
#include<cstring>
#include "external_class.h"

using namespace std;

// TODO also create a Struct and an Enum with the variables
// The variables have mostly the same name to give the VariableUsageResolver a possibly hard time and uncover more faults.

void printLog(string logId, string message){
    cout << logId << ": " << message << endl;
}

class error;

class ScopeVariables{
    public:
        string varName = "instance_field";
        static string staticVarName;

        int functionX();

        void function1(){
            printLog("func1_impl_this_varName", varName);
            printLog("func1_static_staticVarName", staticVarName);
            for(string varName = "first_loop_local"; varName.size() < 17 ; varName += " "){
                printLog("func1_first_loop_varName", varName);
            }

            for(string varName = "second_loop_local"; varName.size() < 18 ; varName += " "){
                {
                    string varName = "local_in_inner_block";
                    printLog("func1_nested_block_shadowed_local_varName", varName);
                }
                printLog("func1_second_loop_varName", varName);
            }
        }

        void function2(string varName){
            printLog("func2_param_varName", varName);
            printLog("func2_this_varName", this->varName);
            // Initializer-Statement that is allowed in newer C++ versions
            if(string varName = "if_local"; varName.size() > 0){
                printLog("func2_if_varName", varName);
            }

            try {
                throw new error();
            } catch (error* varName) {
                printLog("func2_catch_varName", varName);
            };
            ScopeVariables scopeVariables;
            printLog("func2_instance_varName", scopeVariables.varName);
            printLog("func2_imp_this_varName", varName);

        }

        void function3(string varName){
            ScopeVariables scopeVariables;
            printLog("func3_instance_varName", scopeVariables.varName);

            ExternalClass externalClass;
            printLog("func3_external_instance_varName", externalClass.varName);

        }

        void function4(){
            printLog("func4_static_staticVarName", ScopeVariables::staticVarName);
            printLog("func4_external_staticVarName", ExternalClass::staticVarName);
            ExternalClass externalClass;
            printLog("func4_external_instance_varName", externalClass.varName);
            printLog("func4_second_external_staticVarName", ExternalClass::staticVarName);
        }

        void function5(){
            ScopeVariables first;
            ScopeVariables second;
            first.staticVarName = "staticVarName_Of_Both";

            printLog("func5_staticVarName_throughInstance_first", first.staticVarName);
            printLog("func5_staticVarName_throughInstance_second", second.staticVarName);
        }

        // C++ inner classes are currently not parsed
        class InnerClass {
            public:
                string varName = "inner_instance_field";
                static string staticVarName;

                void function1(){
                    printLog("func1_inner_imp_this_varName", varName);
                    InnerClass inner;
                    ScopeVariables scopeVariables;
                    printLog("func1_inner_instance_varName", inner.varName);
                    printLog("func1_outer_instance_varName", scopeVariables.varName);
                    printLog("func1_outer_static_staticVarName", ScopeVariables::staticVarName);
                    printLog("func1_inner_static_staticVarName", ScopeVariables::InnerClass::staticVarName); // Can i remove the ScopeVariables:: ??
                // There is no special reference to the outer-class like ClassName.this.varname as in Java

                }


                void function2(string varName){
                    string staticVarName = "inner_local_named_static";
                    InnerClass inner;
                    ScopeVariables scopeVariables;
                    printLog("func2_inner_instance_varName_with_shadows", inner.varName);
                    printLog("func2_outer_instance_varName_with_shadows", scopeVariables.varName);
                    printLog("func2_outer_static_staticVarName_with_shadows", ScopeVariables::staticVarName);
                    printLog("func2_inner_static_staticVarName_with_shadows", ScopeVariables::InnerClass::staticVarName); // Can i remove the ScopeVariables:: ??
                // There is no special reference to the outer-class like ClassName.this.varname as in Java

                }

        };
};
int main (int argc, char *argv[]) {
    string varName = "parameter";
    ScopeVariables scopeVariables;
    scopeVariables.function1();
    scopeVariables.function2(varName);
    scopeVariables.function3(varName);
    scopeVariables.function4();
    scopeVariables.function5();
    ScopeVariables::InnerClass innerClass;
    innerClass.function1();
    innerClass.function2(varName);
    printLog("main_local_varName", varName);
}

string ScopeVariables::staticVarName = "static_field";
string ScopeVariables::InnerClass::staticVarName = "inner_static_field";

int ScopeVariables::functionX(){
    return 0;
}
